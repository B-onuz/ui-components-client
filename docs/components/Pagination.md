# Pagination (Component)

Componente para paginação.

## Como utilizar

```js
import Pagination from 'ui-components/lib/components/Pagination'
```

```jsx
import React, { useState } from 'react'
import Pagination from 'ui-components/lib/components/Pagination'

export default {
  title: 'Pagination',
  component: Pagination,
}

export const PaginationExample = () => {
  const [currentPage, setCurrentPage] = useState(1)
  const total = 18

  const handleChangeNext = () => {
    const nextPage = currentPage + 1
    if (nextPage <= total) {
      setCurrentPage(nextPage)
    }
  }

  const handleChangePrev = () => {
    const prevPage = currentPage - 1
    if (prevPage > 0) {
      setCurrentPage(currentPage - 1)
    }
  }

  return (
    <Pagination
      color={'primary'}
      m={20}
      onChangePage={(item) => setCurrentPage(item)}
      onPrevPage={(item) => handleChangePrev()}
      onNextPage={(item) => handleChangeNext()}
      total={total}
      page={currentPage}
      onLazyChangePage={(page) => console.log(page)}
    />
  )
}
```

## Variações

### Variant

```jsx
<Pagination color={'primary'} />
```

Escala no tema: `theme.colors`

## Styled-system props

- space https://styled-system.com/table#space
- layout https://styled-system.com/table#layout
