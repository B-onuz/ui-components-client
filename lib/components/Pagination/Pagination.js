"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

exports.__esModule = true;
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _styledSystem = require("styled-system");

var _Button = _interopRequireDefault(require("../Button"));

var _this = void 0,
    _jsxFileName = "/Users/master/projects/bonuz/ui-components/src/components/Pagination/Pagination.js";

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject2() {
  var data = _taggedTemplateLiteralLoose(["\n  position: relative;\n  border-radius: 32px;\n  min-width: 32px;\n  height: 32px;\n  padding: 0 0.3em;\n  background: none;\n  overflow: hidden;\n  -webkit-mask-image: -webkit-radial-gradient(white, black);\n  ", ";\n  &::before {\n    content: '';\n    width: 100%;\n    height: 100%;\n    position: absolute;\n    top: 0;\n    right: 0;\n    bottom: 0;\n    left: 0;\n    ", ";\n    transform: translateY(-100%);\n    transition: 0.2s ease-in-out;\n  }\n  &:active {\n    background-color: transparent;\n  }\n  &:hover,\n  &:active {\n    background-color: transparent;\n    &::before {\n      transform: translateY(0);\n      z-index: 1;\n    }\n    color: #fff;\n  }\n  ", ";\n  & > span {\n    position: relative;\n    z-index: 2;\n  }\n  &.ellipsis {\n    &::before {\n      content: none;\n      color: #e0e1e2;\n    }\n    &:hover {\n      cursor: default;\n      color: #e0e1e2;\n      background-color: #f7f8f8;\n    }\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteralLoose(["\n  display: flex;\n  flex-wrap: wrap;\n  ", "\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

var StyledPagination = _styledComponents["default"].div(_templateObject(), (0, _styledSystem.compose)(_styledSystem.layout, _styledSystem.space));

var PageItem = (0, _styledComponents["default"])(_Button["default"])(_templateObject2(), function (_ref) {
  var theme = _ref.theme,
      color = _ref.color;
  return "color: " + theme.colors[color];
}, function (_ref2) {
  var theme = _ref2.theme,
      color = _ref2.color;
  return "background-color: " + theme.colors[color];
}, function (_ref3) {
  var theme = _ref3.theme,
      color = _ref3.color,
      active = _ref3.active;
  return !!active && "\n    color: #fff;\n    background-color: " + theme.colors[color] + ";\n  ";
});
var Pagination = (0, _react.memo)(function (_ref4) {
  var children = _ref4.children,
      color = _ref4.color,
      onChangePage = _ref4.onChangePage,
      onNextPage = _ref4.onNextPage,
      onPrevPage = _ref4.onPrevPage,
      total = _ref4.total,
      page = _ref4.page,
      onLazyChangePage = _ref4.onLazyChangePage,
      _ref4$lazyTimeout = _ref4.lazyTimeout,
      lazyTimeout = _ref4$lazyTimeout === void 0 ? 300 : _ref4$lazyTimeout,
      rest = _objectWithoutPropertiesLoose(_ref4, ["children", "color", "onChangePage", "onNextPage", "onPrevPage", "total", "page", "onLazyChangePage", "lazyTimeout"]);

  var _useState = (0, _react.useState)(),
      lazyKey = _useState[0],
      setLazyKey = _useState[1];

  var _useState2 = (0, _react.useState)(true),
      isFirstItem = _useState2[0],
      setIsFirstItem = _useState2[1];

  var _useState3 = (0, _react.useState)(true),
      isLastItem = _useState3[0],
      setIsLastItem = _useState3[1];

  var _useState4 = (0, _react.useState)([]),
      pageMarkers = _useState4[0],
      setPageMarkers = _useState4[1];

  var defineMarkers = function defineMarkers(input) {
    var marker = 0;
    var markersArray = [];

    while (marker < input) {
      markersArray[marker] = marker + 1;
      marker += 1;
    }

    setPageMarkers(markersArray);

    if (markersArray.length <= 7) {
      setIsFirstItem(true);
      setIsLastItem(true);
    }
  };

  (0, _react.useEffect)(function () {
    defineMarkers(total);
  }, []);
  (0, _react.useEffect)(function () {
    handleEllipsis();
    if (!onLazyChangePage) return;
    window.clearTimeout(lazyKey);
    setLazyKey(setTimeout(function () {
      onLazyChangePage(page);
    }, lazyTimeout));
  }, [page]);

  var handleEllipsis = function handleEllipsis() {
    if (total >= 8) {
      if (page < 5) {
        setIsFirstItem(true);
        setIsLastItem(false);
      } else if (page > total - 4) {
        setIsFirstItem(false);
        setIsLastItem(true);
      } else {
        setIsFirstItem(false);
        setIsLastItem(false);
      }
    }
  };

  var handleMarkers = function handleMarkers() {
    if (!!isFirstItem || !!isLastItem) {
      if (!!isFirstItem && !!isLastItem) {
        return pageMarkers.map(function (item, index) {
          return /*#__PURE__*/_react["default"].createElement(PageItem, {
            color: color,
            active: page === item,
            m: 1,
            key: index,
            onClick: function onClick() {
              return onChangePage(item);
            },
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 123,
              columnNumber: 11
            }
          }, /*#__PURE__*/_react["default"].createElement("span", {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 124,
              columnNumber: 13
            }
          }, item));
        });
      } else if (!!isFirstItem) {
        return pageMarkers.map(function (item, index) {
          while (index <= 4) {
            return /*#__PURE__*/_react["default"].createElement(PageItem, {
              color: color,
              active: page === item,
              m: 1,
              key: index,
              onClick: function onClick() {
                return onChangePage(item);
              },
              __self: _this,
              __source: {
                fileName: _jsxFileName,
                lineNumber: 131,
                columnNumber: 15
              }
            }, /*#__PURE__*/_react["default"].createElement("span", {
              __self: _this,
              __source: {
                fileName: _jsxFileName,
                lineNumber: 132,
                columnNumber: 17
              }
            }, item));
          }

          if (index === 7) return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(PageItem, {
            color: 'default',
            disabled: true,
            m: 1,
            className: "ellipsis",
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 138,
              columnNumber: 17
            }
          }, /*#__PURE__*/_react["default"].createElement("span", {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 139,
              columnNumber: 19
            }
          }, '...')), /*#__PURE__*/_react["default"].createElement(PageItem, {
            color: color,
            m: 1,
            key: total,
            onClick: function onClick() {
              return onChangePage(total);
            },
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 141,
              columnNumber: 17
            }
          }, /*#__PURE__*/_react["default"].createElement("span", {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 142,
              columnNumber: 19
            }
          }, total)));
        });
      } else if (!!isLastItem) {
        return pageMarkers.map(function (item, index) {
          if (index === 0) return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(PageItem, {
            color: color,
            m: 1,
            key: 1,
            onClick: function onClick() {
              return onChangePage(1);
            },
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 152,
              columnNumber: 17
            }
          }, /*#__PURE__*/_react["default"].createElement("span", {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 153,
              columnNumber: 19
            }
          }, "1")), /*#__PURE__*/_react["default"].createElement(PageItem, {
            color: 'default',
            m: 1,
            className: "ellipsis",
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 155,
              columnNumber: 17
            }
          }, /*#__PURE__*/_react["default"].createElement("span", {
            __self: _this,
            __source: {
              fileName: _jsxFileName,
              lineNumber: 156,
              columnNumber: 19
            }
          }, '...')));

          while (total <= index + 5) {
            return /*#__PURE__*/_react["default"].createElement(PageItem, {
              color: color,
              active: page === item,
              m: 1,
              key: index,
              onClick: function onClick() {
                return onChangePage(item);
              },
              __self: _this,
              __source: {
                fileName: _jsxFileName,
                lineNumber: 162,
                columnNumber: 15
              }
            }, /*#__PURE__*/_react["default"].createElement("span", {
              __self: _this,
              __source: {
                fileName: _jsxFileName,
                lineNumber: 163,
                columnNumber: 17
              }
            }, item));
          }
        });
      }
    } else {
      return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement(PageItem, {
        color: color,
        m: 1,
        key: 1,
        onClick: function onClick() {
          return onChangePage(1);
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 171,
          columnNumber: 11
        }
      }, /*#__PURE__*/_react["default"].createElement("span", {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 172,
          columnNumber: 13
        }
      }, "1")), /*#__PURE__*/_react["default"].createElement(PageItem, {
        color: 'default',
        m: 1,
        className: "ellipsis",
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 174,
          columnNumber: 11
        }
      }, /*#__PURE__*/_react["default"].createElement("span", {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 175,
          columnNumber: 13
        }
      }, '...')), pageMarkers.map(function (item, index) {
        if (item === page - 1 || item === page || item === page + 1) return /*#__PURE__*/_react["default"].createElement(PageItem, {
          color: color,
          active: page === item,
          m: 1,
          key: index,
          onClick: function onClick() {
            return onChangePage(item);
          },
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 180,
            columnNumber: 17
          }
        }, /*#__PURE__*/_react["default"].createElement("span", {
          __self: _this,
          __source: {
            fileName: _jsxFileName,
            lineNumber: 181,
            columnNumber: 19
          }
        }, item));
      }), /*#__PURE__*/_react["default"].createElement(PageItem, {
        color: 'default',
        disabled: true,
        m: 1,
        className: "ellipsis",
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 185,
          columnNumber: 11
        }
      }, /*#__PURE__*/_react["default"].createElement("span", {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 186,
          columnNumber: 13
        }
      }, '...')), /*#__PURE__*/_react["default"].createElement(PageItem, {
        color: color,
        m: 1,
        key: total,
        onClick: function onClick() {
          return onChangePage(total);
        },
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 188,
          columnNumber: 11
        }
      }, /*#__PURE__*/_react["default"].createElement("span", {
        __self: _this,
        __source: {
          fileName: _jsxFileName,
          lineNumber: 189,
          columnNumber: 13
        }
      }, total)));
    }
  };

  return /*#__PURE__*/_react["default"].createElement(StyledPagination, _extends({}, rest, {
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 196,
      columnNumber: 10
    }
  }), handleMarkers());
});
var _default = Pagination;
exports["default"] = _default;
module.exports = exports.default;