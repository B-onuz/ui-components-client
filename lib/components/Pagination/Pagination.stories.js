'use strict'

function _typeof(obj) {
  '@babel/helpers - typeof'
  if (typeof Symbol === 'function' && typeof Symbol.iterator === 'symbol') {
    _typeof = function _typeof(obj) {
      return typeof obj
    }
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === 'function' && obj.constructor === Symbol && obj !== Symbol.prototype ? 'symbol' : typeof obj
    }
  }
  return _typeof(obj)
}

exports.__esModule = true
exports.PaginationExample = exports['default'] = void 0

var _react = _interopRequireWildcard(require('react'))

var _Pagination = _interopRequireDefault(require('./Pagination'))

var _this = void 0,
  _jsxFileName = '/home/junior/Fontes/ui-components-client/src/components/Pagination/Pagination.stories.js'

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj }
}

function _getRequireWildcardCache() {
  if (typeof WeakMap !== 'function') return null
  var cache = new WeakMap()
  _getRequireWildcardCache = function _getRequireWildcardCache() {
    return cache
  }
  return cache
}

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj
  }
  if (obj === null || (_typeof(obj) !== 'object' && typeof obj !== 'function')) {
    return { default: obj }
  }
  var cache = _getRequireWildcardCache()
  if (cache && cache.has(obj)) {
    return cache.get(obj)
  }
  var newObj = {}
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor
  for (var key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc)
      } else {
        newObj[key] = obj[key]
      }
    }
  }
  newObj['default'] = obj
  if (cache) {
    cache.set(obj, newObj)
  }
  return newObj
}

var _default = {
  title: 'Pagination',
  component: _Pagination['default'],
}
exports['default'] = _default

var PaginationExample = function PaginationExample() {
  var _useState = (0, _react.useState)(1),
    currentPage = _useState[0],
    setCurrentPage = _useState[1]

  var total = 7

  var handleChangeNext = function handleChangeNext() {
    var nextPage = currentPage + 1

    if (nextPage <= total) {
      setCurrentPage(nextPage)
    }
  }

  var handleChangePrev = function handleChangePrev() {
    var prevPage = currentPage - 1

    if (prevPage > 0) {
      setCurrentPage(currentPage - 1)
    }
  }

  return /*#__PURE__*/ _react['default'].createElement(_Pagination['default'], {
    color: 'primary',
    m: 20,
    onChangePage: function onChangePage(item) {
      return setCurrentPage(item)
    },
    onPrevPage: function onPrevPage(item) {
      return handleChangePrev()
    },
    onNextPage: function onNextPage(item) {
      return handleChangeNext()
    },
    total: total,
    page: currentPage,
    __self: _this,
    __source: {
      fileName: _jsxFileName,
      lineNumber: 28,
      columnNumber: 5,
    },
  })
}

exports.PaginationExample = PaginationExample
