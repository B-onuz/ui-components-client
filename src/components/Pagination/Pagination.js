import React, { memo, useEffect, useState } from 'react'
import styled from 'styled-components'
import { layout, space, compose } from 'styled-system'
import Button from '../Button'

const StyledPagination = styled.div`
  display: flex;
  flex-wrap: wrap;
  ${compose(layout, space)}
`
const PageItem = styled(Button)`
  position: relative;
  border-radius: 32px;
  min-width: 32px;
  height: 32px;
  padding: 0 0.3em;
  background: none;
  overflow: hidden;
  -webkit-mask-image: -webkit-radial-gradient(white, black);
  ${({ theme, color }) => `color: ${theme.colors[color]}`};
  &::before {
    content: '';
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    ${({ theme, color }) => `background-color: ${theme.colors[color]}`};
    transform: translateY(-100%);
    transition: 0.2s ease-in-out;
  }
  &:active {
    background-color: transparent;
  }
  &:hover,
  &:active {
    background-color: transparent;
    &::before {
      transform: translateY(0);
      z-index: 1;
    }
    color: #fff;
  }
  ${({ theme, color, active }) =>
    !!active &&
    `
    color: #fff;
    background-color: ${theme.colors[color]};
  `};
  & > span {
    position: relative;
    z-index: 2;
  }
  &.ellipsis {
    &::before {
      content: none;
      color: #e0e1e2;
    }
    &:hover {
      cursor: default;
      color: #e0e1e2;
      background-color: #f7f8f8;
    }
  }
`

const Pagination = memo(({ children, color, onChangePage, onNextPage, onPrevPage, total, page, onLazyChangePage, lazyTimeout = 300, ...rest }) => {
  const [lazyKey, setLazyKey] = useState()
  const [isFirstItem, setIsFirstItem] = useState(true)
  const [isLastItem, setIsLastItem] = useState(true)
  const [pageMarkers, setPageMarkers] = useState([])

  const defineMarkers = (input) => {
    let marker = 0
    let markersArray = []
    while (marker < input) {
      markersArray[marker] = marker + 1
      marker += 1
    }
    setPageMarkers(markersArray)
    if (markersArray.length <= 7) {
      setIsFirstItem(true)
      setIsLastItem(true)
    }
  }

  useEffect(() => {
    defineMarkers(total)
  }, [])

  useEffect(() => {
    handleEllipsis()
    if (!onLazyChangePage) return
    window.clearTimeout(lazyKey)
    setLazyKey(
      setTimeout(() => {
        onLazyChangePage(page)
      }, lazyTimeout)
    )
  }, [page])

  const handleEllipsis = () => {
    if (total >= 8) {
      if (page < 5) {
        setIsFirstItem(true)
        setIsLastItem(false)
      } else if (page > total - 4) {
        setIsFirstItem(false)
        setIsLastItem(true)
      } else {
        setIsFirstItem(false)
        setIsLastItem(false)
      }
    }
  }

  const handleMarkers = () => {
    if (!!isFirstItem || !!isLastItem) {
      if (!!isFirstItem && !!isLastItem) {
        return pageMarkers.map((item, index) => (
          <PageItem color={color} active={page === item} m={1} key={index} onClick={() => onChangePage(item)}>
            <span>{item}</span>
          </PageItem>
        ))
      } else if (!!isFirstItem) {
        return pageMarkers.map((item, index) => {
          while (index <= 4)
            return (
              <PageItem color={color} active={page === item} m={1} key={index} onClick={() => onChangePage(item)}>
                <span>{item}</span>
              </PageItem>
            )
          if (index === 7)
            return (
              <>
                <PageItem color={'default'} disabled m={1} className="ellipsis">
                  <span>{'...'}</span>
                </PageItem>
                <PageItem color={color} m={1} key={total} onClick={() => onChangePage(total)}>
                  <span>{total}</span>
                </PageItem>
              </>
            )
        })
      } else if (!!isLastItem) {
        return pageMarkers.map((item, index) => {
          if (index === 0)
            return (
              <>
                <PageItem color={color} m={1} key={1} onClick={() => onChangePage(1)}>
                  <span>1</span>
                </PageItem>
                <PageItem color={'default'} m={1} className="ellipsis">
                  <span>{'...'}</span>
                </PageItem>
              </>
            )
          while (total <= index + 5)
            return (
              <PageItem color={color} active={page === item} m={1} key={index} onClick={() => onChangePage(item)}>
                <span>{item}</span>
              </PageItem>
            )
        })
      }
    } else {
      return (
        <>
          <PageItem color={color} m={1} key={1} onClick={() => onChangePage(1)}>
            <span>1</span>
          </PageItem>
          <PageItem color={'default'} m={1} className="ellipsis">
            <span>{'...'}</span>
          </PageItem>
          {pageMarkers.map((item, index) => {
            if (item === page - 1 || item === page || item === page + 1)
              return (
                <PageItem color={color} active={page === item} m={1} key={index} onClick={() => onChangePage(item)}>
                  <span>{item}</span>
                </PageItem>
              )
          })}
          <PageItem color={'default'} disabled m={1} className="ellipsis">
            <span>{'...'}</span>
          </PageItem>
          <PageItem color={color} m={1} key={total} onClick={() => onChangePage(total)}>
            <span>{total}</span>
          </PageItem>
        </>
      )
    }
  }

  return <StyledPagination {...rest}>{handleMarkers()}</StyledPagination>
})
export default Pagination
