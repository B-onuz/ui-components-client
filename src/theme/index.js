const colors = {
  red: '#db2828',
  lightRed: '#ef4d5a',
  mediumRed: '#FF4F5D',
  darkRed: '#9c2543',
  fullDarkRed: '#550e33',
  orange: '#f2711c',
  yellow: '#fbbd08',
  olive: '#b5cc18',
  green: '#21BA45',
  teal: '#00b5ad',
  blue: '#2185d0',
  facebookBlue: '#3B5998',
  blueLink: '#4285F4',
  violet: '#6435c9',
  purple: '#a333c8',
  pink: '#e03997',
  brown: '#a5673f',
  grey: '#767676',
  textGrey: '#8D8D8D',
  lightGrey: '#6D6D6D',
  mediumGrey: '#838383',
  darkGrey: '#9A9A9A',
  boldGray: '#55555A',
  black: '#1b1c1d',
  white: '#fdfdfd',
  warning: '#7A4D05',
  success: '#1E561F',
  info: '#0E566C',
  danger: '#973937',
  cyan: '#1CB5AD',
  disabledGrey: '#ffffff1f',
  disabledGreyText: '#ffffff4d',
}

colors.primary = colors.lightRed
colors.secondary = colors.fullDarkRed
colors.default = '#e0e1e2'

const breakpoints = ['40em', '52em', '64em', '80em']

breakpoints.sm = breakpoints[0]
breakpoints.md = breakpoints[1]
breakpoints.lg = breakpoints[2]
breakpoints.xl = breakpoints[3]

const fontSizes = [12, 14, 16, 20, 24, 32]

export default {
  fontFamily: `'Lato', sans-serif`,
  colors,
  fontSizes,
  breakpoints,
}
