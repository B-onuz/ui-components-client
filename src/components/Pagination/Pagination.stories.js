import React, { useState } from 'react'
import Pagination from './Pagination'

export default {
  title: 'Pagination',
  component: Pagination,
}

export const PaginationExample = () => {
  const [currentPage, setCurrentPage] = useState(1)
  const total = 7

  const handleChangeNext = () => {
    const nextPage = currentPage + 1
    if (nextPage <= total) {
      setCurrentPage(nextPage)
    }
  }

  const handleChangePrev = () => {
    const prevPage = currentPage - 1
    if (prevPage > 0) {
      setCurrentPage(currentPage - 1)
    }
  }

  return (
    <Pagination
      color={'primary'}
      m={20}
      onChangePage={(item) => setCurrentPage(item)}
      onPrevPage={(item) => handleChangePrev()}
      onNextPage={(item) => handleChangeNext()}
      total={total}
      page={currentPage}
    />
  )
}
