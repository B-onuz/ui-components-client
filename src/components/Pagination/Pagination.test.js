import React from 'react'
import Pagination from './Pagination.js'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import theme from '../../theme'
import 'jest-styled-components'

test('when render Pagination', () => {
  const component = renderer.create(
    <ThemeProvider theme={theme}>
      <Pagination
        color={'primary'}
        m={20}
        onChangePage={() => {}}
        onPrevPage={() => {}}
        onNextPage={() => {}}
        total={20}
        page={11}
        onLazyChangePage={() => {}}
      />
    </ThemeProvider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
