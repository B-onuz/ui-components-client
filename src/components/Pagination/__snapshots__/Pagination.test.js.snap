// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when render Pagination 1`] = `
.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  font-family: inherit;
  border-radius: 4px;
  outline: none;
  cursor: pointer;
  padding: .5rem 1rem;
  border-style: solid;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  border-color: undefined !important;
  color: #fff;
  background-color: #ef4d5a;
  border-color: #ef4d5a;
  font-size: 16px;
  margin: 4px;
  box-sizing: border-box;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:focus {
  box-shadow: 0 0 8px #ef4d5a;
}

.c1:active:not(:disabled) {
  box-shadow: inset 0px 0px 5px #c1c1c1;
}

.c1:active,
.c1:hover {
  background-color: hsl(355.20000000000005,91.9%,55.8%);
}

.c1:disabled {
  background-color: hsl(355.20000000000005,50.1%,68.2%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin: 20px;
}

.c2 {
  position: relative;
  border-radius: 32px;
  min-width: 32px;
  height: 32px;
  padding: 0 0.3em;
  background: none;
  overflow: hidden;
  -webkit-mask-image: -webkit-radial-gradient(white,black);
  color: #ef4d5a;
}

.c2::before {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #ef4d5a;
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c2:active {
  background-color: transparent;
}

.c2:hover,
.c2:active {
  background-color: transparent;
  color: #fff;
}

.c2:hover::before,
.c2:active::before {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
  z-index: 1;
}

.c2 > span {
  position: relative;
  z-index: 2;
}

.c3 {
  position: relative;
  border-radius: 32px;
  min-width: 32px;
  height: 32px;
  padding: 0 0.3em;
  background: none;
  overflow: hidden;
  -webkit-mask-image: -webkit-radial-gradient(white,black);
  color: #ef4d5a;
  color: #fff;
  background-color: #ef4d5a;
}

.c3::before {
  content: '';
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  background-color: #ef4d5a;
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c3:active {
  background-color: transparent;
}

.c3:hover,
.c3:active {
  background-color: transparent;
  color: #fff;
}

.c3:hover::before,
.c3:active::before {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
  z-index: 1;
}

.c3 > span {
  position: relative;
  z-index: 2;
}

<div
  className="c0"
>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      &lt;
    </span>
  </button>
  <button
    className="c1 c3"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      11
    </span>
  </button>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      12
    </span>
  </button>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      13
    </span>
  </button>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      14
    </span>
  </button>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      15
    </span>
  </button>
  <button
    className="c1 c2"
    color="primary"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <span>
      &gt;
    </span>
  </button>
</div>
`;
